### ISSUE TITLE
Everything Uploaded To S3 Is Content-Type application/octet-stream

### ISSUE THREAD

**Issue by mydigia** [s3fs-fuse/s3fs-fuse#1217]
## My environment info:

$ s3fs --version
```
Amazon Simple Storage Service File System V1.85(commit:unknown) with OpenSSL
Copyright (C) 2010 Randy Rizun <rrizun@gmail.com>
License GPL2: GNU GPL version 2 <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

#### Version of fuse being used (pkg-config --modversion fuse, rpm -qi fuse, dpkg -s fuse)
$ pkg-config --modversion fuse
`2.9.2`

$ rpm -qi fuse, dpkg -s fuse
```
package fuse, is not installed
package dpkg is not installed
Name        : fuse
Version     : 2.9.2
Release     : 11.el7
Architecture: x86_64
Install Date: Mon 16 Dec 2019 15:51:58 UTC
Group       : System Environment/Base
Size        : 223297
License     : GPL+
Signature   : RSA/SHA256, Mon 12 Nov 2018 14:25:34 UTC, Key ID 24c6a8a7f4a80eb5
Source RPM  : fuse-2.9.2-11.el7.src.rpm
Build Date  : Tue 30 Oct 2018 21:32:35 UTC
Build Host  : x86-01.bsys.centos.org
Relocations : (not relocatable)
Packager    : CentOS BuildSystem <http://bugs.centos.org>
Vendor      : CentOS
URL         : https://github.com/libfuse/libfuse
Summary     : File System in Userspace (FUSE) utilities
Description :
With FUSE it is possible to implement a fully functional filesystem in a
userspace program. This package contains the FUSE userspace tools to
mount a FUSE filesystem.
normal        /etc/fuse.conf
normal        /usr/bin/fusermount
normal        /usr/bin/ulockmgr_server
normal        /usr/sbin/mount.fuse
normal        /usr/share/doc/fuse-2.9.2
normal        /usr/share/doc/fuse-2.9.2/AUTHORS
normal        /usr/share/doc/fuse-2.9.2/COPYING
normal        /usr/share/doc/fuse-2.9.2/ChangeLog
normal        /usr/share/doc/fuse-2.9.2/FAQ
normal        /usr/share/doc/fuse-2.9.2/Filesystems
normal        /usr/share/doc/fuse-2.9.2/NEWS
normal        /usr/share/doc/fuse-2.9.2/README
normal        /usr/share/doc/fuse-2.9.2/README.NFS
normal        /usr/share/man/man1/fusermount.1.gz
normal        /usr/share/man/man1/ulockmgr_server.1.gz
normal        /usr/share/man/man8/mount.fuse.8.gz
```

#### Kernel information (uname -r)
```
$ uname -r
3.10.0-1062.9.1.el7.x86_64
```

#### GNU/Linux Distribution, if applicable (cat /etc/os-release)
```
$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

#### s3fs command line used, if applicable
```
s3fs BUCKET_NAME /FOLDER_NAME -o use_cache=/tmp -o allow_other -o uid=1000 -o mp_umask=002 -o multireq_max=5 -o use_path_request_style -o url=https://s3-eu-west-2.amazonaws.com -o nonempty
```

### Details about issue

When I mount my s3 bucket and then use rsync to sync my data to the S3 bucket, then every single object on the bucket has the following metadata:

```
Content-Type | application/octet-stream
  | x-amz-meta-ctime | 1576596267
  | x-amz-meta-gid | 1000
  | x-amz-meta-mode | 33188
  | x-amz-meta-mtime | 1439986331
  | x-amz-meta-uid | 1000
```

It looks like "Content-Type | application/octet-stream" stop my S3 static website hosting from working, because it downloads the files instead of serving them.

If I upload a single object using Cyberduck to S3 the metadata is correctly picked and served on the public URL of the bucket.

**Comment by gaul** [comment]
@mydigia I wrote a test for this behavior and s3fs correctly detects text/plain and image/jpeg.  Could you be more specific about which file types are not detected?

**Comment by nickmilleruk** [comment]
I am experiencing exactly the same issue. Here are the details of my installation:

$ s3fs --version

```
Amazon Simple Storage Service File System V1.86 (commit:unknown) with OpenSSL
Copyright (C) 2010 Randy Rizun <rrizun@gmail.com>
License GPL2: GNU GPL version 2 <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

$ rpm -qi fuse
```
Name        : fuse
Version     : 2.9.2
Release     : 11.el7
Architecture: x86_64
Install Date: Sun 15 Dec 2019 10:42:25 GMT
Group       : System Environment/Base
Size        : 223297
License     : GPL+
Signature   : RSA/SHA256, Mon 12 Nov 2018 14:25:34 GMT, Key ID 24c6a8a7f4a80eb5
Source RPM  : fuse-2.9.2-11.el7.src.rpm
Build Date  : Tue 30 Oct 2018 21:32:35 GMT
Build Host  : x86-01.bsys.centos.org
Relocations : (not relocatable)
Packager    : CentOS BuildSystem <http://bugs.centos.org>
Vendor      : CentOS
URL         : https://github.com/libfuse/libfuse
Summary     : File System in Userspace (FUSE) utilities
Description :
With FUSE it is possible to implement a fully functional filesystem in a
userspace program. This package contains the FUSE userspace tools to
mount a FUSE filesystem.
```
$ uname -r
`3.10.0-1062.12.1.el7.x86_64`

$ cat /etc/os-release
`NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"`

I am mounting my S3 bucket as follows:

`s3fs#s3.bucket.name /path/to/mount/to _netdev,allow_other,mp_umask=022,use_path_request_style,url=https://s3.us-east-1.amazonaws.com 0 0`

All files uploaded have ContentType set to: application/octet-stream

I have tested with images (.jpg), html files (.html) and text files (.txt).

What additional information do you need to be able to debug this issue?

Thanks
Nick

**Comment by tryggve** [comment]
We had the same issue that all files uploaded got application/octet-stream. It turns out that the ec2-instance running our ftp-server did not have mimetypes configured on the OS level.

https://wiki.debian.org/MIME/etc/mime.types

Once Linux knew how to recognize different data types the ContentType header in s3 looked as we wanted.

**Comment by nickmilleruk** [comment]
@tryggve thank you for this. This was exactly the issue.

This was a new install of Centos 8, and there was no /etc/mime.types file.

The fix was simple - obtain a valid mime.types file (in my case I took one from an existing Centos 7 install) and copy it into place in /etc. Make sure that the file is owned by root (chown root:root /etc/mime.types) and that all users can read it (chmod a+r /etc/mime.types).

Thanks again for the pointer.

**Comment by juliogonzalez** [comment]
Is this happening with s3fs-fuse compiled from source, or also with the s3fuse-packages from EPEL?

> in my case I took one from an existing Centos 7 install

Looks to me that maybe you are missing the package that provides it.

This is from a Docker container:
```
[root@5b6a4438d9c5 /]# cat /etc/os-release 
NAME="CentOS Linux"
VERSION="8 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="CentOS Linux 8 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:8"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-8"
CENTOS_MANTISBT_PROJECT_VERSION="8"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="8"
```
```
[root@5b6a4438d9c5 /]# ls /etc/mime.types
ls: cannot access '/etc/mime.types': No such file or directory
```
```
[root@5b6a4438d9c5 /]# yum install -y mailcap
Failed to set locale, defaulting to C
Last metadata expiration check: 0:00:19 ago on Wed Mar 18 19:52:14 2020.
Dependencies resolved.
==================================================================================================================================================================================
 Package                                   Arch                                     Version                                        Repository                                Size
==================================================================================================================================================================================
Installing:
 mailcap                                   noarch                                   2.1.48-3.el8                                   BaseOS                                    39 k

Transaction Summary
==================================================================================================================================================================================
Install  1 Package

Total download size: 39 k
Installed size: 71 k
Downloading Packages:
mailcap-2.1.48-3.el8.noarch.rpm                                                                                                                    36 kB/s |  39 kB     00:01    
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                              16 kB/s |  39 kB     00:02     
warning: /var/cache/dnf/BaseOS-f6a80ba95cf937f2/packages/mailcap-2.1.48-3.el8.noarch.rpm: Header V3 RSA/SHA256 Signature, key ID 8483c65d: NOKEY
CentOS-8 - Base                                                                                                                                   1.6 MB/s | 1.6 kB     00:00    
Importing GPG key 0x8483C65D:
 Userid     : "CentOS (CentOS Official Signing Key) <security@centos.org>"
 Fingerprint: 99DB 70FA E1D7 CE22 7FB6 4882 05B5 55B3 8483 C65D
 From       : /etc/pki/rpm-gpg/RPM-GPG-KEY-centosofficial
Key imported successfully
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                          1/1 
  Installing       : mailcap-2.1.48-3.el8.noarch                                                                                                                              1/1 
  Verifying        : mailcap-2.1.48-3.el8.noarch                                                                                                                              1/1 

Installed:
  mailcap-2.1.48-3.el8.noarch                                                                                                                                                     

Complete!
```
```
[root@5b6a4438d9c5 /]# rpm -qf /etc/mime.types
mailcap-2.1.48-3.el8.noarch
```

If it required for s3fs-fuse to work, maybe we should make sure the file exists, and require mailcap for the RPMs/DEBs, etc.

Could be that for most user this works because the mailcap is installed by default on standard setups, but maybe on minimal installations this is not the case and it should be required.

**Comment by juliogonzalez** [comment]
Well, in fact we already have mailcap at the CentOS instructions:
https://github.com/s3fs-fuse/s3fs-fuse/wiki/Installation-Notes#fedora--centos--rhel

But no mention to the package or `/etc/mime.types` on the generic instructions.

And it seems I removed it from the spec at 1.84:

https://github.com/juliogonzalez/s3fs-fuse-rpm/blob/master/SPECS/s3fs-fuse.spec#L89

Because it is NOT a build requirement, but a runtime requirement.

So two things need a fix:
a) The `COMPILE.md` file to mention this somehow (so it's generic enough)
b) The SPEC for EPEL/Fedora (link above, will take care of it)
c) The SPEC for SLE/openSUSE (https://build.opensuse.org/package/view_file/filesystems/s3fs/s3fs.spec?expand=1 can handle it as well).

**Comment by juliogonzalez** [comment]
For my own reference, `c` is NOT needed. On SLE/openSUSE `/etc/mime.types` is part of `aaa_base` that is **always installed** even of the most minimal setups.

On Debian/Ubuntu is part of `mime-support` which is **not** installed on the Docker images, and I guess maybe it's absent on minimal installations, so yet another reason to mention it at `COMPILE.md`.

However the Debian package is right, as it requires `mime-support`.

**Comment by juliogonzalez** [comment]
Bodhi requests with the fixes for:

Fedora30: https://bodhi.fedoraproject.org/updates/FEDORA-2020-951f18a723
Fedora31: https://bodhi.fedoraproject.org/updates/FEDORA-2020-92af351313
Fedora32: https://bodhi.fedoraproject.org/updates/FEDORA-2020-3aaa1f559c
EPEL7: https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2020-e6bada7ce3
EPEL8: https://bodhi.fedoraproject.org/updates/FEDORA-EPEL-2020-932011492d
Rawhide: No bodhi enablement, it will be released automatically from the koji build (https://koji.fedoraproject.org/koji/taskinfo?taskID=42608070)

**Comment by nickmilleruk** [comment]
@juliogonzalez thank you for following up on this.

I can confirm the package was installed from EPEL, not source. Also, mailcap is indeed not installed, and this was a minimal install of Centos.

I does look like this should be a required dependency for s3fs-fuse.

It is great to see such a rapid and complete response to an issue like this from an open-source project. Thank you.

Nick

**Comment by ggtakec** [comment]
I looked at this issue and comment of #1253 and I created #1254.
It is that Linux and macOS have a default mime.types path, and this file path can be specified by "mime" option.
And if mime.types cannot be loaded correctly, s3fs will fail to start.

Please let me know if you find any problems, I think the PR will be merged soon.

**Comment by ggtakec** [comment]
I closed this issue because #1254 had been merged into the current master.
If you still have problems, please reopen.
Thanks,
